File: data/readable_format/fortran/file864.f90
module foo_mod
  implicit none
  private
  public :: foo, foo2

  type :: foo
    integer :: i
  contains
    procedure, pass :: doit
    procedure, pass :: getit
  end type foo

  type, extends(foo) :: foo2
    integer :: j
  contains
    procedure, pass :: doit => doit2
    procedure, pass :: getit => getit2
  end type foo2

contains
  subroutine doit(this)
    class(foo), intent(inout) :: this
    this%i = 1
  end subroutine doit

  function getit(this) result(res)
    class(foo), intent(in) :: this
    integer :: res
    res = this%i
  end function getit

  subroutine doit2(this)
    class(foo2), intent(inout) :: this
    this%i = 2
    this%j = 3
  end subroutine doit2

  function getit2(this) result(res)
    class(foo2), intent(in) :: this
    integer :: res
    res = this%j
  end function getit2

end module foo_mod

module bar_mod
  use foo_mod, only: foo
  implicit none
  private
  public :: bar

  type :: bar
    class(foo), allocatable :: a
  contains
    procedure, pass :: doit => doitbar
    procedure, pass :: getit => getitbar
  end type bar

contains
  subroutine doitbar(this)
    class(bar), intent(inout) :: this
    call this%a%doit()
  end subroutine doitbar

  function getitbar(this) result(res)
    class(bar), intent(in) :: this
    integer :: res
    res = this%a%getit()
  end function getitbar
end module bar_mod

program test_foo
  use foo_mod
  use bar_mod
  implicit none

  type(bar) :: b
  integer :: result

  ! Test for base class functionality
  allocate(foo :: b%a)
  call b%doit()
  result = b%getit()
  if (result /= 1) then
    print *, "Test failed for base class. Expected 1, got", result
    stop
  end if
  deallocate(b%a)

  ! Test for derived class functionality
  allocate(foo2 :: b%a)
  call b%doit()
  result = b%getit()
  if (result /= 3) then
    print *, "Test failed for derived class. Expected 3, got", result
    stop
  end if

  print *, "All tests passed."

end program test_foo

File: data/readable_format/cpp/file864.cpp
#include <iostream>
#include <memory>
#include <cstdlib>

// Foo class definition
class Foo {
public:
    virtual void doit() { i = 1; }
    virtual int getit() const { return i; }
protected:
    int i{0};
};

// Foo2 class definition, derived from Foo
class Foo2 : public Foo {
public:
    void doit() override {
        i = 2;
        j = 3;
    }
    int getit() const override { return j; }
private:
    int j{0};
};

// Bar class definition
class Bar {
public:
    void doit() { a->doit(); }
    int getit() { return a->getit(); }
    template <typename T>
    void allocate() {
        a = std::make_unique<T>();
    }
private:
    std::unique_ptr<Foo> a;
};

// Main program with tests
int main() {
    Bar a;

    // Test with base class functionality
    a.allocate<Foo>();
    a.doit();
    if (a.getit() != 1) {
        std::cerr << "Test failed for base class. Expected 1, got " << a.getit() << std::endl;
        std::abort();
    }

    // Test with derived class functionality
    a.allocate<Foo2>();
    a.doit();
    if (a.getit() != 3) {
        std::cerr << "Test failed for derived class. Expected 3, got " << a.getit() << std::endl;
        std::abort();
    }

    std::cout << "All tests passed." << std::endl;
    return 0;
}